service/matching/version_sets.go

```diff
@@ -253,6 +253,30 @@ func updateImpl(timestamp hlc.Clock, existingData *persistencespb.VersioningData
 			BuildIds: buildIDsCopy,
 		}
 		makeVersionInSetDefault(&modifiedData, targetSetIdx, versionInSetIdx, &timestamp)
+	} else if mergeSets := req.GetMergeSets(); mergeSets != nil {
+		if targetSetIdx == -1 {
+			return nil, serviceerror.NewNotFound(fmt.Sprintf("targeted primary version %v not found", targetedVersion))
+		}
+		secondaryBuildID := mergeSets.GetSecondarySetBuildId()
+		secondarySetIdx, _ := findVersion(&modifiedData, secondaryBuildID)
+		if secondarySetIdx == -1 {
+			return nil, serviceerror.NewNotFound(fmt.Sprintf("targeted secondary version %v not found", secondaryBuildID))
+		}
+		if targetSetIdx == secondarySetIdx {
+			// Nothing to be done
+			return existingData, nil
+		}
+		// Merge the sets together, preserving the primary set's default by making it have to most recent timestamp.
+		primarySet := modifiedData.VersionSets[targetSetIdx]
+		primarySet.DefaultUpdateTimestamp = &timestamp
```

Comment:
Don't mutate the original, copy the set first.

Reply:
This is the "modified" data, which is already a copy. Why would we need to copy it again?

Reply:
Ah ok nevermind I see the initial copy is still keeping pointers to all the internal sets